### ISSUE TITLE
Different distance per each lens + "Match DoF" button

### ISSUE THREAD

**Issue by Gabriele-LS** [patik/dof#35]
Hi @patik, I tested the new GUI and found it fantastic.

What about allowing the user to set a different distance from the subject for each lens?

What about having a checkbox to define the reference lens and a "Match DoF" button that automatically adjusts the aperture of all the lenses so that their DoF is the same as the one of the reference lens?

I do a lot of 2- and 3-camera interviews, and I struggle to have the same background on all cameras.

Sample setup:

![3-Cam-Setup](https://user-images.githubusercontent.com/67373152/210227891-e2deef98-c31b-4925-8acd-4b2404ec2824.jpg)

Source: https://www.premiumbeat.com/blog/interview-with-three-cameras/

Thanks!

**Comment by patik** [comment]
Hey @Gabriele-LS thanks for the idea! I love hearing how people use the app.

I'm still thinking about how this could be achieved. Here's what I've come up with so far.

1. Add a distance field to each row (as you suggested)
2. Make the DoF editable

Recall that DoF, focal length, crop factor, and aperture all go together—when you change one of them, one of the other properties must change. So the question arises: when I change one value, which of the other values should be set automatically?

![CleanShot 2023-01-03 at 19 46 19@2x](https://user-images.githubusercontent.com/262137/210421644-5a0be339-969a-470b-82d0-a26f98cbacf3.png)

For the interface to make sense, I think that at least one of the columns needs to become 'locked' or read-only. That is, if you want to automatically recalculate the aperture while you're changing the DoF, then the focal length needs to be locked. 

![CleanShot 2023-01-03 at 19 49 55@2x](https://user-images.githubusercontent.com/262137/210422449-ee9e8cfc-e538-46b9-8d99-534ef7ba9471.png)

Or, if you wanted to know how set the aperture while setup a certain distance from the subject, you would lock the focal length, then manually adjust the DoF and see the aperture change automatically. (I believe it's safe to say that we can usually lock the sensor column since that's not easily swappable like a lens.) Does this make sense?

This raises another question—should the locking apply to the same column for every lens? (That is, if the user wants to lock the DoF, then it'll be locked for all lenses.) If so, that's not too difficult to resolve from a design standpoint. But if not—if you want each lens to have its own independent set of characteristics, including which column is locked—then that feels like it could be an entirely different use case than what the table is currently meant to do, so it could mean implementing a whole different interface.

What do you think?

**Comment by Gabriele-LS** [comment]
Hello @patik, thank you very much for evaluating my proposal.

Regarding my use case, here is what usually happens on set.

1. I arrive at the location with my three cameras. Thus, the first certain datum is the **size of the sensors**.
2. I mount a lens on each camera, position each camera around the interviewee, and frame the various shots. Once if finish this step, I have another two certain datums: the **focal length** I chose for each lens and the **distance** between each camera and the subject. 
3. I need to have consistent DoF across all the shots. So I expect the calculator to give me the aperture I need to set on each camera (given its sensor size, focal length, and distance from the subject). I think this calculation is called something like "reverse DoF calculation."

Nota bene: Once reached step 3 I will never change any data resulting from steps 1 and 2. For instance, I will never change my focal length or the distance from the subject just to achieve a consistent DoF (this would change the framing of my shots).

Long story short: I need to be able to enter the sensor size, the focal length, and the subject distance for each camera, and I need the calculator to tell me what aperture I have to use on each camera.

Does that make sense?

**Comment by patik** [comment]
Hi @Gabriele-LS. I've been working on how to calculate the aperture when the depth of field is known. It seems that I would actually need two values: the DoF, plus either the **near limit** or **far limit** of the DoF.

I tried to guess these limits based on the DoF and distance to subject. That is, if we assume that the subject is perfectly in the middle of the DoF, then we can calculate `near limit = distance - (DoF / 2)`. However, this seems to produce results that are off by 1-2 f-stops for many data sets.

I see that you have two use cases:

1. Calculate the DoF for one lens/camera, then apply this to the other lenses to get their aperture
2. Manually type in a DoF, then determine the aperture for that lens

I _think_ I can satisfy the first use case because I can just copy the near/far limit values from the first lens to the other lenses. However, this may be violating some laws of physics 🙂  Is it even possible for two different lenses (different sensor sizes, different focal lengths) to have the exact same DoF, near limit, and far limit? 🤔

I cannot support the second use case unless I also ask for either the near or far limit, so you would have to know what the value is.

Thoughts?

**Comment by Gabriele-LS** [comment]
Hi @patik, thanks for your update.

I think the first use case perfectly matches the multi-camera interview situation. It is not relevant to have the exact same depth of field. It is enough to have a similar result (similar background blur) among the 2-3 cameras used. In my opinion, the process should be as follows:

1. The user enters the distance from the subject, the focal length, the sensor size, and the aperture for the first camera.
2. The user enters the distance from the subject, the focal length, and the sensor size for the second camera.
3. The user clicks a "Match Camera 1's DoF" button (for instance) next to the aperture menu for the second camera, and the software automatically chooses the correct aperture needed to achieve a similar background blur.

What do you think?
